### DISCUSSION TITLE
snapshots: Does waitForSelector happen after the JS in execute()?

### DISCUSSION THREAD

**Discussion by thecristen** [percy/cli#538] [Support] [answered]
Hi! Thanks for the great service, but I'm trying to keep up here and I have a question.
I have a snapshot configured as follows:

```js
{
  name: ...,
  url: ...,
  execute() {
      // open each of the collapsed results
      const accordions = document.querySelectorAll('.c-accordion-ui__trigger')
      for (const a of accordions) { a.click(); }
  },
  waitForSelector: ".customClass"
}
```

How my page works is that, after one of those collapsed elements are clicked, it triggers a chain of events that eventually results in a `customClass` class name being added to a DOM element. I'd like to wait for that before I take the snapshot.

When I run @percy/cli to take snapshot it reports an error that it failed to find `.customClass`.

I'm curious as to whether this should work as described, or if this is user error or what.

**Comment by Robdel12** [comment, marked as answer]
Hey @thecristen! We haven't documented this yet (it's quite new and we've been occupied with a couple networking bugs since #513). We pass a `waitFor` function to execute, so you can make sure the page state is what you'd expect before executing your JavaScript. For example:

```js
module.exports = async () => {
  return [
    {
      name: 'example 1',
      url: 'https://percy.io',
      execute: {
        async beforeResize({ waitFor }) {
          await waitFor(() => !!document.querySelector('.hero-main-image'), 5000);
        }
      }
    }
  ];
};
```

Or using your code:

```js
[{
  name: '...',
  url: '...',
  async execute({ waitFor }) {
      // open each of the collapsed results
      await waitFor(() => !!document.querySelector('.customClass'), 5000);
      const accordions = document.querySelectorAll('.c-accordion-ui__trigger')
      for (const a of accordions) { a.click(); }
  },
}]
```

**Reply by wwilsman** [reply]
I wanted to add an explanation here for what @thecristen is seeing. the `execute` function is synonymous with `execute.beforeSnapshot`, which will be executed _just before the snapshot is taken_. The `waitForSelector` option will wait for a selector _before_ `execute.beforeSnapshot` is called.

It looks like @thecristen wants to wait for an element _after_ `execute.beforeSnapshot`. Since we can't possibly cover all "wait for" scenarios, a helper is provided to `execute` functions for this exact reason. Although what @thecristen is looking for would probably be using `waitFor` after clicking the accordions:

```js
async execute({ waitFor }) {
  const accordions = document.querySelectorAll('.c-accordion-ui__trigger');
  for (const a of accordions) a.click();

  // wait for any .customClass element to appear within 5000ms
  await waitFor(() => !!document.querySelector('.customClass'), 5000);
}
```
